src/dataflow/src/render/join/delta_join.rs

```diff
@@ -22,6 +32,132 @@ use crate::operator::CollectionExt;
 use crate::render::datum_vec::DatumVec;
 use crate::render::join::{JoinBuildState, JoinClosure};
 
+/// A delta query is implemented by a set of paths, one for each input.
+///
+/// Each delta query path responds to its input changes by repeated lookups
+/// in arrangements for other join inputs. These lookups require specific
+/// instructions about which expressions to use as keys. Along the way,
+/// various closures are applied to filter and project as early as possible.
+pub struct DeltaJoinPlan {
+    /// The set of path plans.
+    ///
+    /// Each path identifies its source relation, so the order is only
+    /// important for determinism of dataflow construction.
+    path_plans: Vec<DeltaPathPlan>,
+}
+
+/// A delta query path is implemented by a sequences of stages,
+pub struct DeltaPathPlan {
+    /// The relation whose updates seed the dataflow path.
+    source_relation: usize,
+    /// An initial closure to apply before any stages.
+    initial_closure: JoinClosure,
+    /// A *sequence* of stages to apply one after the other.
+    stage_plans: Vec<DeltaStagePlan>,
+    /// A concluding closure to apply after the last stage.
+    final_closure: JoinClosure,
+}
+
+/// A delta query stage performs a stream lookup into an arrangement.
+pub struct DeltaStagePlan {
+    /// The relation index into which we will look up.
+    lookup_relation: usize,
+    /// The key expressions to use for the stream source.
```

Comment:
haven't read the rest of the code yet but: the comments here reference the `stream source` which I think is the same thing as the `source_relation` in the `DeltaPathPlan` object but the overloading of `source` makes things a little challenging to understand. I think it would help a lot to use `relation` wherever we are talking about data, and `source` and `lookup` as the two relations we may be referring to? I think we could rename `stream_keys` -> `source_keys` and it would be more obvious whats happening

Reply:
100% agree!

Reply:
Wait, I changed my mind! Sorry. There is a source relation, but after it has been joined with a lookup relation, I wouldn't call it the "source" any more. I think that is the reason for the different names. Lemme double check to see if that is consistent though. Or, if that (once explained in comments) resolves things great.